---
title: "Tn-seq mapping QC"
output: html_document
date: "2024-04-12"
params:
  quantif_rds: "data/fastq/fastq/Mm.ht2.tn.quantif/robj.rds"
  genome_dir: "data/ref/Mm"
---

```{r,message=FALSE,include=FALSE}
library(Rsamtools)
library(SummarizedExperiment)
library(tidyverse)
library(reshape2)
library(rtracklayer)
library(GenomicRanges)
```

```{r echo=FALSE}
x <- readRDS(params$quantif_rds)
gff <- import.gff(file.path(params$genome_dir,"genome.gff.gz"),feature.type="gene")

# Add overlapping genes
h <- findOverlaps(rowRanges(x),GRanges(gff),type="within")
rowData(x)$gene_IDs <- extractList(gff$ID,as(h,"IntegerList"))


# head(gff)

x$num_inserted_TA <- colSums(assays(x)$counts>0)
x$num_in_TA <- colSums(assays(x)$counts)

# Helper function to extract a data.frame of all gene-expression values
# suitable to display a heatmap
makeAssayDF <- function(x,assay_idx="counts",col_meta=colnames(colData(x)),row_meta=character(0)) {
  assay(x,assay_idx) |>
    as.data.frame() |>
    rownames_to_column("row_id") |>
    pivot_longer(!row_id,names_to = "col_id") |>
    full_join(as(colData(x)[col_meta],"data.frame") |> rownames_to_column("col_id"),by="col_id") |> # Add all colData
    full_join(as(rowData(x)[row_meta],"data.frame") |> rownames_to_column("row_id"),by="row_id") |> # Add all rowData
    mutate(row_id=factor(row_id,rownames(x)),col_id=factor(col_id,colnames(x)))
}


```


### Number of inserted TA in each library
```{r echo=FALSE}
colData(x) |>
  as("data.frame")|>
  ggplot(aes(x=num_inserted_TA,y=lib))+
  geom_col()

```

### Number of insertion divided by inserted TA  in each library
```{r echo=FALSE}

colData(x) |>
  as("data.frame")|>
  ggplot(aes(x=num_in_TA/num_inserted_TA,y=lib))+
  geom_col()
```

### Number of insertion in each TA to compare library

```{r echo=FALSE}
# Show TA pairs
K <- makeAssayDF(x)
full_join(K,K,by="row_id",relationship = "many-to-many") |>
  filter(value.x>0,value.y>0) |>
  filter(lib.x>lib.y) |>
  ggplot(aes(x=value.x,y=value.y)) + 
    facet_grid(lib.y~lib.x) +
    geom_point()
```

### TA number and TA inserted for each gene

```{r echo=FALSE}
makeAssayDF(x,row_meta = "gene_IDs") |>
  unnest(gene_IDs) |>
  group_by(gene_IDs,col_id) |>
  summarise(n_TA=n(),n_inserted=sum(value>0)) |>
  ggplot(aes(x=1+n_TA,y=1+n_inserted)) + 
    facet_wrap(~col_id) +
    geom_point() +
    scale_x_log10() + scale_y_log10()
  




# data.frame(idx=queryHits(h),ID=gff$ID[subjectHits(h)]) |>
#  group_by(idx)|>
#  summarise(list(ID))

# as.list(h)



#nb de TA avec insertion (% nb de ta total)
#nb de TA avec insertion graph bar pour les echantillons

#nb de ta par gene


#charger gff et annot√© ta avec les genes (find overlap)
#nb de ta avec insertion par gene
#nb d'insertion pour 10 ta
#scater plot
```


```{r eval=FALSE,echo=FALSE}
rowData(x)
rowRanges(x)

gr <- GRanges(assay_counts$X)
mcols(gr)$fwd <- assay_fwd$BBB017_R1_cutadapt.Mm.ht2.bam
mcols(gr)$fwd_01 <- assay_fwd$BBB019_R1_cutadapt.Mm.ht2.bam



assays(x)

dim(x)

class(rowData(x))

colData(x)
metadata(x)
assays(x)
assays(x)$counts
assays(x)[1:3,1:3]
assays(x,'fwd')
assay(x,'fwd')
names(assays(x))
mcols(x)

rowRanges(x)
rowData(x)

a <- as.data.frame(assays(x)$counts)
a$gene <- rownames(a)

a[a$BBB017_R1_cutadapt.Mm.ht2.bam<=10 & a$BBB019_R1_cutadapt.Mm.ht2.bam<=10,]



```

```{r eval=FALSE,echo=FALSE}
a <- as.data.frame(assays(x)$counts)
a$gene <- rownames(a)
b <- a$gene
a <- melt(a,id='gene')

b <- b[1:2]

ggplot(a,aes(gene,value,fill=variable))+
  geom_col(position = 'dodge')+
  scale_x_discrete(limits=b)+
  ylim(0,100)



```



