---
title: "Genome QC"
output: html_document
date: "2024-04-12"
params:
  genome_dir: "data/ref/Dd+Mm"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,warning = FALSE, message = FALSE)
options(dplyr.print_max = 1e3)
```

```{r echo=FALSE, message=FALSE}
library(tidyverse)
library(Rsamtools)
```


```{r message=FALSE}
contigs_stats <- function(fa) {
	alphabetFrequency(getSeq(fa)) |> 
			as_tibble() |>
			select(where(~any(.>0))) |>
			mutate(contig = seqlevels(fa)) |>
			full_join(enframe(seqlengths(fa),"contig","size"),by="contig") |>
			relocate(contig,size)
}
```

# Load data

Name of the analysed genome: `r params$genome_dir`

```{r}
fa <- FaFile(file.path(params$genome_dir,"genome.fasta"))
gtf <- rtracklayer::import(file.path(params$genome_dir,"genome.gtf.gz"),genome=seqinfo(fa))

if (file.exists(file.path(params$genome_dir,"seqinfo.tsv"))) {
	si <- read.table(file.path(params$genome_dir,"seqinfo.tsv"),sep="\t",header=TRUE)
	genome(gtf) <- si$genome[match(seqlevels(gtf),si$contig)]
}
```

# Sequence Statistics {.tabset}
## Genome-wide {.tabset}

```{r}
ctg_stat <- contigs_stats(fa) |>
	mutate(genome=genome(gtf)) |>
	relocate(genome)

dinuc_stat <- dinucleotideFrequency(getSeq(fa)) |>
	as_tibble() |>
	mutate(contig = seqlevels(fa),genome=genome(gtf)) |>
	relocate(genome,contig)
```

### Nucleotide content {.tabset}

#### Graphical Overview
```{r}
ctg_stat |>
	group_by(genome) |> 
	select(!size) |>
	pivot_longer(!c(genome,contig),names_to = "nucleotide",values_to = "count") |>
	group_by(genome,nucleotide) |>
	summarise(count=sum(count)) |>
	ggplot(aes(x=count,y=genome,fill=nucleotide)) + 
		geom_col(position = "fill",color="black") + 
		geom_text(aes(label=nucleotide),position = position_fill(vjust=0.5)) +
		theme_bw() + xlab("") + 
		scale_x_continuous(labels = scales::percent) + theme(legend.position = "none")
```

#### Summary
```{r}
ctg_stat  |>
	group_by(genome) |> 
	summarise(across(!contig,sum),num_contig=n()) %>%
	mutate(GC_content = 100*rowSums(pick(any_of(c("C","G"))))/size) |>
	mutate(GC_content = sprintf("%.1f%%",GC_content)) |>
	relocate(genome,size,num_contig,GC_content) |>
	knitr::kable(caption="Number of nucleotide per genome")
```

### Dinucleotide content {.tabset}

#### Graphical Overview
```{r fig.height=4, fig.width=7}
dinuc_stat |>
	pivot_longer(!c(genome,contig),names_to = "dinucleotide",values_to = "count") |>
	group_by(genome,dinucleotide) |>
	summarize(count=sum(count)) |>
	group_by(genome) |>
	mutate(pct=100*count/sum(count)) |>
	ggplot(aes(y=dinucleotide,x=count)) + 
		facet_grid(.~genome,scales="free_x") +
		geom_col() + 
		geom_text(aes(x=count/2,label=sprintf("%.1f%%",pct)),size=3) + 
		theme_bw() + theme(legend.position="none")
```


#### Summary
```{r}
dinuc_stat |>
	group_by(genome) |>
	summarize(across(!contig,sum)) |>
	knitr::kable()
```





## Contig-wide {.tabset}

### Nucleotide content {.tabset}

#### Graphical Overview
```{r fig.height=8, fig.width=6}
ctg_stat |>
	select(!size) |>
	pivot_longer(!c(genome,contig),names_to = "nucleotide",values_to = "count") |>
	ggplot(aes(x=count,y=contig,fill=nucleotide)) + 
		facet_grid(genome~.,scales = "free_y",space = "free_y") +
		geom_col(position = "fill",color="black") + 
		geom_text(aes(label=nucleotide),position = position_fill(vjust=0.5)) +
		theme_bw() + xlab("") + 
		scale_x_continuous(labels = scales::percent) + theme(legend.position = "none")
```

#### Summary
```{r}
ctg_stat |>
	arrange(genome,desc(size)) |>
	mutate(GC_content = 100*rowSums(pick(any_of(c("C","G"))))/size) |>
	mutate(GC_content = sprintf("%.1f%%",GC_content)) |>
	relocate(genome,contig,size,GC_content) |>
	knitr::kable(caption="Contigs Nucleotide Content")
```


### Dinucleotide content {.tabset}

#### Graphical Overview
```{r fig.height=4, fig.width=8}
dinuc_stat |>
	pivot_longer(!c(genome,contig),names_to = "dinucleotide",values_to = "count") |>
	ggplot(aes(y=dinucleotide,x=count)) + 
	facet_grid(.~genome + contig,scales="free_x") +
	geom_col() + 
	geom_text(aes(x=count/2,label=count),size=3) + 
	theme_bw() + theme(legend.position="none")
```


#### Summary
```{r}
dinuc_stat |>
	knitr::kable()
```


# Features statistics {.tabset}

```{r}
genes <- subset(gtf,type %in% "gene")
genes$dna <- getSeq(fa,genes)
genes$dinuc <- dinucleotideFrequency(genes$dna)
```

## Feature Count {.tabset}

### Global

```{r}
feature_type_cov <- split(gtf,gtf$type) |>
	map(~sum(width(reduce(split(.,seqnames(.)),ignore.strand=TRUE)))) |>
	simplify2array() |>
	as_tibble(rownames="contig") |>
	pivot_longer(!contig,names_to = "feature_type",values_to = "bp_cov")

feature_count <- as.data.frame(gtf) |>
	select(contig=seqnames,feature_type=type,feature_size=width) |>
	left_join(select(ctg_stat,genome,contig),by="contig") |>
	group_by(genome,feature_type) |>
	summarize(n_feature=n(),median_feature_size=median(feature_size),avg_feature_size=mean(feature_size)) 

pivot_wider(feature_count,id_cols = "genome",names_from = "feature_type",values_from = "n_feature") |>
	knitr::kable(caption="Number of feature in the genome")
```

### By Contig
```{r}
as.data.frame(gtf) |>
	mutate(genome=genome(gtf)[seqnames]) |>
	group_by(genome,seqnames,type) |>
	summarize(n=n()) |>
	pivot_wider(names_from = "type",values_from = "n") |>
	arrange(genome,desc(gene)) |>
	knitr::kable(caption = "Number of feature per contig")
```

## Features Size {.tabset}

### Distribution
```{r}
as.data.frame(gtf) |>
	ggplot(aes(x=width)) +
	facet_wrap(~type) +
	geom_histogram() + 
	scale_x_log10() + theme_bw() + xlab("Feature size") + ylab("Feature count")
```

### Summary
```{r}
pivot_wider(feature_count,id_cols = "genome",names_from = "feature_type",values_from = "avg_feature_size") |>
	knitr::kable(caption="Average feature size")

pivot_wider(feature_count,id_cols = "genome",names_from = "feature_type",values_from = "median_feature_size") |>
	knitr::kable(caption="Median feature size")
```




## Covered Genome Fraction
```{r}
select(ctg_stat,genome,contig,contig_size=size) |>
	full_join(feature_type_cov,by="contig") |>
	group_by(genome,feature_type) |>
	summarize(covered_bp=sum(bp_cov)) |>
	pivot_wider(id_cols = "genome",names_from = "feature_type",values_from = "covered_bp") |>
	knitr::kable(caption="Number of base-pair covered by all features of a given type")
```

## Genes Dinucleotide Content {.tabset}

Here we report only features of type `gene`.

### Distribution
```{r fig.height=4, fig.width=15}
as_tibble(genes$dinuc) |>
	mutate(genome = unname(genome(genes))[as.integer(seqnames(genes))]) |>
	pivot_longer(!genome,names_to = "dinucleotide") |>
	ggplot() +
		facet_grid(genome~dinucleotide,scales="free_y") +
		geom_histogram(aes(x=pmin(value,100))) + 
		ylab("Gene count") + xlab("Dinucleotide count") + 
		ggtitle("Number of gene for a given number of dinucleotide")
```

### Summary
```{r}
as_tibble(genes$dinuc) |>
	mutate(genome = unname(genome(genes))[as.integer(seqnames(genes))]) |>
	pivot_longer(!genome,names_to = "dinucleotide") |>
	group_by(genome,dinucleotide) |>
	summarise(
		num_gene = n(),
		num_gene_with_ge10_dinuc = sum(value>=10),
		avg_dinuc_count = mean(value),
		median_dinuc_count = median(value),
		min_dinuc_count = min(value),
		max_dinuc_count = max(value)
	) |>
	knitr::kable(caption="This table look only at features of type `gene`, and report Average/Median/Min/Max number of dinucleotide per gene. `num_gene_with_ge10_dinuc` is the number of gene with more than 10 occurence of the dinucleotide")
```











