---
title: "Differential expression analysis"
output: html_document
author: Julien Prados
date: "2023-08-03"
params:
  bam_dir: "data/fastq/test"
  genome_dir: "data/ref/DdMm"
  outdir: "./"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```


```{r,message=FALSE,include=FALSE}
library(DESeq2)
#library(fgsea)
#library(igraph)
source("src/lib_hisat2.R")
```


# Load quantification and add genes metadata
```{r,echo=TRUE,include=FALSE}
# Load read counts
x <- file.path(params$bam_dir) |>
	list.files(".ht2.bam$",recursive=TRUE,full.names=TRUE) |>
	ht2_read_feature_counts()

# Assign genome associated to each contig
genome(metadata(x)$seqinfo) <- ifelse(seqlevels(metadata(x)$seqinfo) %in% c("NC_010612.1","NC_010604.1"),"Mm","Dd")

# Retreive gene_biotype and contig from GTF
gtf <- rtracklayer::import(
	file.path(params$genome_dir,"genome.gtf"),
	genome = metadata(x)$seqinfo,
	feature.type="gene")
rowData(x)$gene_biotype <- gtf$gene_biotype[match(rowData(x)$feature_id,gtf$gene_id)]
rowData(x)$contig <- as.character(seqnames(gtf))[match(rowData(x)$feature_id,gtf$gene_id)]
```



# Add samples metadata
```{r,warning=FALSE}
# Simplify library names
x$lib <- sub("\\.[^\\.]*\\.ht2\\.bam","",x$lib)

# Define conditions to compare
x$cond <- sample(c("condA","condB"),size = ncol(x),replace = TRUE)

# Alternatively load this metadata from an excel file 
# colData(x) <- left_join(
# 	as.data.frame(colData(x)),
# 	readxl::read_xlsx("data/libraries_metadata.xlsx"),
# 	by = "lib",relationship="one-to-one"
# )
```


# Apply DESeq pipeline
```{r}
x <- DESeqDataSet(x,design= ~ cond) |>
  DESeq()

resultsNames(x)
```

# PCA + HEATMAP





# Volcano
```{r}
results(x) |>
  as.data.frame() |>
  rownames_to_column("gene") |>
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point() +
    geom_vline(xintercept = c(-1, 1),linetype = "dashed") +
    geom_hline(yintercept = -log10(0.05),linetype = "dashed") +
    ggrepel::geom_text_repel(aes(label=gene), max.overlaps=10,data=\(x) filter(x,abs(log2FoldChange)>1 & padj<0.05)) +
    theme_bw()
```




# DESeq PCA and Heatmap sample distance
```{r}
X <- rlogTransformation(x)
plotPCA(X,"cond")
```


```{r}
assay(X) <- assay(X) - rowMeans(assay(X))
X <- sample(X,1000)
X <- X[hclust(dist(assay(X)),method="ward.D2")$order]

assay(X) |>
  as.data.frame() |>
  rownames_to_column("gene_id") |>
  mutate(gene_id=factor(gene_id,gene_id)) |>
  pivot_longer(!gene_id) |>
    ggplot() + 
      geom_tile(aes(x=gene_id,y=name,fill=value)) + 
      scale_fill_gradient2() + 
      theme_void() + 
      theme(axis.text.y = element_text(),panel.border = element_rect(fill=NA))
heatmap(assay(X),Rowv = as.dendrogram(h),scale="none",col=colorRampPalette(c("blue","white","red"))(10))
```








