---
title: "Differential expression analysis"
output: html_document
date: "2023-08-03"
params:
  bam_dir: "data/fastq/pilot"
  genome_dir: "data/ref/DdMm"
  outdir: "./"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```


```{r,message=FALSE,include=FALSE}
library(Rsamtools)
library(SummarizedExperiment)
library(DESeq2)
library(tidyverse)
#library(fgsea)
#library(igraph)
```


# Load quantification and add genes metadata
```{r,echo=TRUE,include=FALSE}
# Load read counts
x <- readRDS("out/data/DdMm/robj.rds")

# Assign genome associated to each contig
genome(metadata(x)$seqinfo) <- case_when(
	seqlevels(metadata(x)$seqinfo) %in% seqlevels(FaFile("data/ref/Dd/genome.fasta")) ~ "Dd",
	seqlevels(metadata(x)$seqinfo) %in% seqlevels(FaFile("data/ref/Mm/genome.fasta")) ~ "Mm",
	TRUE ~ NA
)

# Retreive gene_biotype and contig from GTF
gtf <- rtracklayer::import(
	file.path(params$genome_dir,"genome.gtf"),
	genome = metadata(x)$seqinfo,
	feature.type="gene")
rowData(x) <- left_join(
	as.data.frame(rowData(x)),
	as.data.frame(gtf) |> select(feature_id=gene_id,contig=seqnames,gene_biotype,old_locus_tag,gene),
	by="feature_id",relationship="one-to-one")
```



# Add samples metadata
```{r,warning=FALSE}
# Simplify library names
colnames(x) <- x$lib <- sub("_S._L..._.*","",x$lib)

# Define conditions to compare
x$cond <- local({set.seed(4321);sample(c("condA","condB","condC"),size = ncol(x),replace = TRUE)})

# Alternatively load this metadata from an excel file 
# colData(x) <- left_join(
# 	as.data.frame(colData(x)),
# 	readxl::read_xlsx("data/libraries_metadata.xlsx"),
# 	by = "lib",relationship="one-to-one"
# )
```


# Apply DESeq pipeline
```{r}
x <- DESeqDataSet(x,design= ~ cond) |>
  DESeq()

all_deseq_results <- list(
	condAvsB = as.data.frame(results(x,contrast = c("cond","condA","condB"))),
	condAvsC = as.data.frame(results(x,contrast = c("cond","condA","condC")))
) |>
	map(~rownames_to_column(as.data.frame(.),"feature_id")) |>
  bind_rows(.id = "comparison") |>
  mutate(direction = case_when(
    log2FoldChange >= +1 & padj <= 0.05 ~ "up",
    log2FoldChange <= -1 & padj <= 0.05 ~ "down",
    TRUE ~ "ns"
  ))
```


# Volcano plot
```{r}
all_deseq_results |>
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
		facet_wrap(~comparison) +
    geom_point(aes(color=direction)) +
    geom_vline(xintercept = c(-1, 1),linetype = "dashed") +
    geom_hline(yintercept = -log10(0.05),linetype = "dashed") +
    ggrepel::geom_text_repel(aes(label=feature_id), max.overlaps=15,data=\(x) filter(x,!direction%in%"ns"),size=3) +
    theme_bw()
```




# PCA and Heatmap
```{r}
X <- rlogTransformation(x)
plotPCA(X,"cond")
```


```{r}
# Heatmap of 1000 randomly selected genes
heatmap(assay(sample(X,1000)),scale="row",col=colorRampPalette(c("blue","white","red"))(11))
```








